File: Data Structures/hashing/dart/hash.dart
void main() {
  // Create a hash map
  var hashMap = <String, int>{};

  // Add key-value pairs to the hash map
  hashMap['apple'] = 1;
  hashMap['banana'] = 2;
  hashMap['cherry'] = 3;

  // Print the hash map
  print(hashMap); // {apple: 1, banana: 2, cherry: 3}

  // Access a value using a key
  print(hashMap['banana']); // 2

  // Update a value using a key
  hashMap['banana'] = 4;
  print(hashMap); // {apple: 1, banana: 4, cherry: 3}

  // Remove a key-value pair
  hashMap.remove('banana');
  print(hashMap); // {apple: 1, cherry: 3}

  // Check if a key exists
  print(hashMap.containsKey('apple')); // true
  print(hashMap.containsKey('banana')); // false

  // Check if a value exists
  print(hashMap.containsValue(1)); // true
  print(hashMap.containsValue(2)); // false

  // Get the number of key-value pairs
  print(hashMap.length); // 2

  // Get all keys
  print(hashMap.keys); // (apple, cherry)

  // Get all values
  print(hashMap.values); // (1, 3)

  // Clear the hash map
  hashMap.clear();
  print(hashMap); // {

  // Create a hash set
  var hashSet = <int>{};

  // Add elements to the hash set
  hashSet.add(1);
  hashSet.add(2);
  hashSet.add(3);

  // Print the hash set
  print(hashSet); // {1, 2, 3}

  // Check if an element exists
  print(hashSet.contains(2)); // true
  print(hashSet.contains(4)); // false

  // Remove an element
  hashSet.remove(2);
  print(hashSet); // {1, 3}

  // Get the number of elements
  print(hashSet.length); // 2

  // Clear the hash set
  hashSet.clear();
  print(hashSet); // {}
}

File: Data Structures/hashing/ts/hash.ts
// Create a hash map
const hashMap = new Map<string, number>();

// Add key-value pairs to the hash map
hashMap.set("apple", 1);
hashMap.set("banana", 2);
hashMap.set("cherry", 3);

// Print the hash map
console.log(hashMap); // Map(3) { 'apple' => 1, 'banana' => 2, 'cherry' => 3 }

// Access a value using a key
console.log(hashMap.get("banana")); // 2

// Update a value using a key
hashMap.set("banana", 4);
console.log(hashMap); // Map(3) { 'apple' => 1, 'banana' => 4, 'cherry' => 3 }

// Remove a key-value pair
hashMap.delete("banana");
console.log(hashMap); // Map(2) { 'apple' => 1, 'cherry' => 3 }

// Check if a key exists
console.log(hashMap.has("apple")); // true
console.log(hashMap.has("banana")); // false

// Check if a value exists
// Map doesn't have a direct containsValue method, so we iterate
console.log([...hashMap.values()].includes(1)); // true
console.log([...hashMap.values()].includes(2)); // false

// Get the number of key-value pairs
console.log(hashMap.size); // 2

// Get all keys
console.log(hashMap.keys()); // [Map Iterator] { 'apple', 'cherry' }

// Get all values
console.log(hashMap.values()); // [Map Iterator] { 1, 3 }

// Clear the hash map
hashMap.clear();
console.log(hashMap); // Map(0) {}

// Create a hash set
const hashSet = new Set<number>();

// Add elements to the hash set
hashSet.add(1);
hashSet.add(2);
hashSet.add(3);

// Print the hash set
console.log(hashSet); // Set(3) { 1, 2, 3 }

// Check if an element exists
console.log(hashSet.has(2)); // true
console.log(hashSet.has(4)); // false

// Remove an element
hashSet.delete(2);
console.log(hashSet); // Set(2) { 1, 3 }

// Get the number of elements
console.log(hashSet.size); // 2

// Clear the hash set
hashSet.clear();
console.log(hashSet); // Set(0) {}
